modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -6361,6 +6830,235 @@ bool GDScriptAnalyzer::class_exists(const StringName &p_class) const {
+					// Copied over trait function.
+					p_class->add_member(trait_member.function);
+					if (trait_member.function->is_static && !p_class->has_static_data) {
+						p_class->has_static_data = true;
+					}
+					break;
+				case GDScriptParser::ClassNode::Member::ENUM:
+					// Copied over trait enum.
+					p_class->add_member(trait_member.m_enum);
+					break;
+				case GDScriptParser::ClassNode::Member::ENUM_VALUE:
+					// Copied over trait enum_values.
+					p_class->add_member(trait_member.enum_value);
+					break;
+				case GDScriptParser::ClassNode::Member::SIGNAL:
+					// Copied over trait signal.
+					p_class->add_member(trait_member.signal);
+					break;
+				case GDScriptParser::ClassNode::Member::GROUP:
+					// Copied over trait annotation.
+					p_class->add_member_group(trait_member.annotation);
+					break;
+				default:
+					continue;
+			}
+
+			// Add trait origin to copied over members.
+			if (!trait_member.get_source_node()->trait_origin.has(p_trait->fqcn)) {
+				trait_member.get_source_node()->trait_origin.append(p_trait->fqcn);
+			}
+			continue;
+		}
+
+		GDScriptParser::ClassNode::Member class_member = p_class->get_member(trait_member.get_name());
+		if (class_member.type != trait_member.type) {
+			push_error(vformat(R"("%s" %s shadows %s in trait "%s".)", class_member.get_name(), class_member.get_type_name(), trait_member.get_type_name(), p_trait_name), class_member.get_source_node());
+			continue;
+		}
+
+		// Skip already copied members by other traits used.
+		if (trait_member.get_source_node()->trait_origin.has(p_class->fqcn)) {
+			continue;
+		}
+		bool copied_over = false;
+		for (const String &trait_fqcn : p_class->traits_fqtn) {
+			if (trait_member.get_source_node()->trait_origin.has(trait_fqcn)) {
+				copied_over = true;
+				break;
+			}
+		}
+		if (copied_over) {
+			continue;
+		}
+
+		// Traits are not allowed to extenting or overriding members traits.
+		if (p_class->type == GDScriptParser::Node::TRAIT) {
+			push_error(vformat(R"("%s" %s is already declared in trait "%s". Traits do not extend/override members from using traits.)", trait_member.get_name(), trait_member.get_type_name(), p_trait_name), class_member.get_source_node());
+			continue;
+		}
+
+		// Extending or overriding members.
+		switch (trait_member.type) {
+			case GDScriptParser::ClassNode::Member::FUNCTION: {
+				override_member_function(class_member.function, trait_member.function, p_trait_name);
+				if (trait_member.function->is_static && !p_class->has_static_data) {
+					p_class->has_static_data = true;
+				}
+			} break;
+			case GDScriptParser::ClassNode::Member::GROUP:
+			case GDScriptParser::ClassNode::Member::UNDEFINED:
+				continue;
+			default:
+				push_error(vformat(R"("%s" %s is shadows already declared %s in trait "%s".)", class_member.get_name(), class_member.get_type_name(), trait_member.get_type_name(), p_trait_name), class_member.get_source_node());
```

Comment:
```suggestion
				push_error(vformat(R"("%s" %s conflicts with an already-existing %s in trait "%s".)", class_member.get_name(), class_member.get_type_name(), trait_member.get_type_name(), p_trait_name), class_member.get_source_node());
```
Since this is an error and isn't simply "replacing" the trait's version, I don't think it makes sense to refer to this as "shadowing".